### PR TITLE
DEP: improve message for deprecated private API for `sparse`

### LINKED ISSUE #19241: BUG: utils from scipy.sparse.sputils are not available from scipy.sparse module
### Describe your issue.

It seems that the deprecation warning of the `scipy.sparse.sputils` is incorrect. The utilities are not available from the `scipy.sparse` module.

This might be on purpose but then the deprecation warning is misleading.

### Reproducing Code Example

```python
In [2]: from scipy.sparse.sputils import get_index_dtype
<ipython-input-2-8d76e9dbe0fa>:1: DeprecationWarning: Please use `get_index_dtype` from the `scipy.sparse` namespace, the `scipy.sparse.sputils` namespace is deprecated.
  from scipy.sparse.sputils import get_index_dtype

In [3]: from scipy.sparse import get_index_dtype
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
Cell In[3], line 1
----> 1 from scipy.sparse import get_index_dtype

ImportError: cannot import name 'get_index_dtype' from 'scipy.sparse' (/home/glemaitre/mambaforge/envs/dev/lib/python3.10/site-packages/scipy/sparse/__init__.py)
```

### Error message

```shell
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
Cell In[3], line 1
----> 1 from scipy.sparse import get_index_dtype

ImportError: cannot import name 'get_index_dtype' from 'scipy.sparse' (/home/glemaitre/mambaforge/envs/dev/lib/python3.10/site-packages/scipy/sparse/__init__.py)
```

### SciPy/NumPy/Python version and system information

```shell
In [4]: import scipy

In [5]: scipy.__version__
Out[5]: '1.11.2'
```

### PR DESCRIPTION
**Pull request by hoangble**
#### Reference issue

Solving a part of issue gh-18279
closes #19241
#### What does this implement/fix?

Finish #18279 for `sparse`

### CODE REVIEW

File: scipy/_lib/tests/test_public_api.py [hunk 1]
```diff
@@ -339,13 +339,23 @@ def check_importable(module_name):
...
+                          ('scipy.sparse.spfuncs', None),
+                          ('scipy.sparse.sputils', None),
                           ('scipy.stats.biasedurn', None),
                           ('scipy.stats.kde', None),
                           ('scipy.stats.morestats', None),
                           ('scipy.stats.mstats_basic', 'mstats'),
                           ('scipy.stats.mstats_extras', 'mstats'),
                           ('scipy.stats.mvn', None),
                           ('scipy.stats.stats', None)])
+
```
Comment (j-bowhay) [inline scipy/_lib/tests/test_public_api.py:358]:
```suggestion
```

File: scipy/sparse/bsr.py [hunk 2]
```diff
@@ -34,13 +31,6 @@ def __dir__():
...
-            f"{name}. Try looking in scipy.sparse instead.")
-
-    warnings.warn(f"Please use `{name}` from the `scipy.sparse` namespace, "
-                  "the `scipy.sparse.bsr` namespace is deprecated.",
-                  category=DeprecationWarning, stacklevel=2)
-
-    return getattr(_bsr, name)
+    return _sub_module_deprecation(sub_package="sparse", module="bsr",
+                                   private_modules=["_bsr"], all=__all__,
+                                   attribute=name)
```
Comment (j-bowhay) [inline scipy/sparse/bsr.py:36]: New line at end of file needed

File: scipy/sparse/data.py [hunk 3]
```diff
@@ -21,13 +16,6 @@ def __dir__():
...
-            f"{name}. Try looking in scipy.sparse instead.")
-
-    warnings.warn(f"Please use `{name}` from the `scipy.sparse` namespace, "
-                  "the `scipy.sparse.data` namespace is deprecated.",
-                  category=DeprecationWarning, stacklevel=2)
-
-    return getattr(_data, name)
+    return _sub_module_deprecation(sub_package="sparse", module="data",
+                                   private_modules=["_data"], all=__all__,
+                                   attribute=name)
```
Comment (j-bowhay) [inline scipy/sparse/data.py:21]: new line at end of file

File: scipy/sparse/linalg/eigen.py [hunk 4]
```diff
@@ -2,13 +2,14 @@
...
+from scipy._lib.deprecation import _sub_module_deprecation
 
 
 __all__ = [  # noqa: F822
-    'ArpackError', 'ArpackNoConvergence',
-    'eigs', 'eigsh', 'lobpcg', 'svds', 'arpack', 'test'
+    'ArpackError', 
+    'ArpackNoConvergence',
+    'eigs', 
+    'eigsh', 'lobpcg', 'svds', 'arpack', 'test'
```
Comment (j-bowhay) [inline scipy/sparse/linalg/eigen.py:12]: can you revert the change here please

File: scipy/sparse/linalg/isolve.py [hunk 5]
```diff
@@ -2,14 +2,25 @@
...
+    'gmres',
+    'lgmres', 
+    'lsmr',
+    'lsqr',
+    'minres',
+    'qmr',
+    'tfqmr', 
+    'utils', 
+    'iterative', 
+    'test'
```
Comment (j-bowhay) [inline scipy/sparse/linalg/isolve.py:23]: and here

File: scipy/sparse/linalg/matfuncs.py [hunk 6]
```diff
@@ -2,14 +2,20 @@
...
+    'expm', 
+    'inv', 
+    'solve', 
+    'solve_triangular',
+    'spsolve', 
+    'is_pydata_spmatrix', 
+    'LinearOperator',
+    'UPPER_TRIANGULAR', 
+    'MatrixPowerOperator', 
+    'ProductOperator'
```
Comment (j-bowhay) [inline scipy/sparse/linalg/matfuncs.py:18]: and here

File: scipy/sparse/linalg/eigen.py [hunk 7]
```diff
@@ -19,19 +20,7 @@ def __dir__():
...
-        msg = (f"Please use `{name}` from the `scipy.sparse.linalg` namespace,"
-               " the `scipy.sparse.linalg.eigen` namespace is deprecated.")
-
-    warnings.warn(msg, category=DeprecationWarning, stacklevel=2)
-
-    return getattr(_eigen, name)
+    return _sub_module_deprecation(sub_package="sparse.linalg", module="eigen",
+                                private_modules=["_eigen"], all=__all__,
+                                attribute=name)
+
```
Comment (j-bowhay) [inline scipy/sparse/linalg/eigen.py:26]: one too many lines at end of file!

**Review by j-bowhay — CHANGES_REQUESTED** [review]
Thanks @hoangle96 some minor changes required

**Review by hoangble — COMMENTED** [review]
Making sure changes are compliant with PEP 8. Removing unnecessary changes

File: scipy/sparse/bsr.py (hunk 2 above)
Comment (j-bowhay) [reply scipy/sparse/bsr.py:36]: Still missing an new line

File: scipy/sparse/linalg/isolve.py (hunk 5 above)
Comment (j-bowhay) [reply scipy/sparse/linalg/isolve.py:23]: again please revert this change it makes it harder to tell if anything has been removed

**Review by j-bowhay — CHANGES_REQUESTED** [review]
Thanks @hoangle96 still a few changes required

**Comment by hoangle96** [comment]
@j-bowhay I have updated the changes to be PEP8 complaint and revert the unnecessary changes.

**Review by j-bowhay — APPROVED** [review]
Thanks @hoangle96! Congratulations on your first contribution to SciPy, if you are interested in doing one of the other modules from the issue that would be greatly appreciated!

**Pull request merged**